File: shadowed-config-registries/src/main/java/io/github/legacy_fabric_community/shadowedconfig/registries/DefaultedRegistry.java
package io.github.legacy_fabric_community.shadowedconfig.registries;

import java.util.Random;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import net.minecraft.util.Identifier;

public class DefaultedRegistry<T> extends Registry<T> {
	private final Identifier defaultId;
	private T defaultValue;

	public DefaultedRegistry(String defaultId) {
		this(new Identifier(defaultId));
	}

	public DefaultedRegistry(Identifier defaultId) {
		this.defaultId = defaultId;
	}

	@Override
	public <V extends T> V set(int rawId, Identifier id, V entry) {
		if (this.defaultId.equals(id)) {
			this.defaultValue = entry;
		}

		return super.set(rawId, id, entry);
	}

	@Override
	public int getRawId(@Nullable T entry) {
		int i = super.getRawId(entry);
		return i == -1 ? super.getRawId(this.defaultValue) : i;
	}

	@Override
	@Nonnull
	public Identifier getId(T entry) {
		Identifier identifier = super.getId(entry);
		return identifier == null ? this.defaultId : identifier;
	}

	@Override
	@Nonnull
	public T get(@Nullable Identifier id) {
		T object = super.get(id);
		return object == null ? this.defaultValue : object;
	}

	@Override
	@Nonnull
	public T get(int index) {
		T object = super.get(index);
		return object == null ? this.defaultValue : object;
	}

	@Override
	@Nonnull
	public T getRandom(Random random) {
		T object = super.getRandom(random);
		return object == null ? this.defaultValue : object;
	}

	public Identifier getDefaultId() {
		return this.defaultId;
	}

	public T getDefaultValue() {
		return this.defaultValue;
	}
}

File: shadowed-config-registries/src/main/scala/io/github/legacy_fabric_community/shadowedconfig/registries/DefaultedRegistry.scala
package io.github.legacy_fabric_community.shadowedconfig.registries

import java.util.{Optional, Random}
import net.minecraft.util.Identifier
import org.jetbrains.annotations.{NotNull, Nullable}

class DefaultedRegistry[T](val defaultId: Identifier) extends Registry[T] {
	private var defaultValue = _

	def this(defaultId: String) {
		this(new Identifier(defaultId))
	}

	override def set[V <: T](rawId: Int, id: Identifier, entry: V): V = {
		if (this.defaultId == id) this.defaultValue = entry
		super.set(rawId, id, entry)
	}

	override def getRawId(@Nullable entry: T): Int = {
		val i = super.getRawId(entry)
		if (i == -1) super.getRawId(this.defaultValue)
		else i
	}

	@NotNull override def getId(entry: T): Identifier = {
		val identifier = super.getId(entry)
		if (identifier == null) this.defaultId
		else identifier
	}

	@NotNull override def get(@Nullable id: Identifier): T = {
		val `object` = super.get(id)
		if (`object` == null) this.defaultValue
		else `object`
	}

	@NotNull override def get(index: Int): T = {
		val `object` = super.get(index)
		if (`object` == null) this.defaultValue
		else `object`
	}

	@NotNull override def getRandom(random: Random): Optional[T] = {
		val optional = super.getRandom(random)
		if (optional.isPresent) Optional.of(this.defaultValue)
		else optional
	}

	def getDefaultId: Identifier = this.defaultId

	def getDefaultValue: T = this.defaultValue
}
